### ISSUE TITLE
Checkpoint more often during training

### ISSUE THREAD

**Issue by oliverwm1** [ai2cm/ace#347]
When we run our jobs in preemptible mode, or with a wall clock limit, our training gets interupted. Currently we lose all progress since the end of the previous epoch. In cases where the per-epoch time is large (e.g. larger than 1 hour) this can lead to a lot of wasted compute. Ideally we should be able to checkpoint and restart training partway through the `train_one_epoch` function. Doing this right may be a bit of a challenge—e.g. do we need to serialize state of data loader so that we use same shuffling for remainder of restarted epoch?

Related: we may also want to run validation/inference more than once per epoch. This is particularly relevant for large datasets which we might iterate over fewer times (e.g. concatenated climSST, AMIP and ERA5!).

**Comment by oliverwm1** [comment]
Assuming the order of batches will be the same, could just loop over the whole epoch again, but add a `continue` to skip all the steps until you get to the step when the model was checkpointed. Might be some edge cases that require care for distributed training.

**Comment by brianhenn** [comment]
> do we need to serialize state of data loader so that we use same shuffling for remainder of restarted epoch?

Do we? If the data are shuffled, just starting again from a new data loader should be OK, in the sense that it won't introduce bias. Since we don't set the random seed for the data loader (I think?), we don't have any expectation that the shuffled loader would be reproducible in a way that would be broken by having half-epochs of data loaded.

**Comment by brianhenn** [comment]
If that's right, then the main downside with not bothering to return to the preempted loader state would just be that val and inference would no longer occur at the regular intervals of number of batches seen.

**Comment by oliverwm1** [comment]
Yeah, I agree it should not introduce a bias to restart with a data loader with possibly different shuffling.

> If that's right, then the main downside with not bothering to return to the preempted loader state would just be that val and inference would no longer occur at the regular intervals of number of batches seen.

I think we would want to use the `num_batches_seen` variable which is currently serialized to skip ahead to where the preemption happened (i.e. partway through the training epoch).

**Comment by oliverwm1** [comment]
Resolved by ai2cm/full-model#2019
